### DISCUSSION TITLE
Auto3dseg - algorithm bundle vs monai bundle confusion - easy way to convert between the two?

### DISCUSSION THREAD

**Discussion by mea02jct** [Project-MONAI/MONAILabel#1630] [General]
Hi,

I'm trying to standardise our processes for creating and deploying segmentation tools. One potential choice is to use auto3dseg for everything. According to some of the documentation I've read this has the added advantage that trained models should already be in MONAI bundle format (another good standard to work to).

However, although auto3dseg uses algorithm bundles (templates) it seems that these are quite different to 'standard' MONAI bundles. I personally find this quite confusing!

Firstly, I wondered whether different terminology should be used to avoid this mis-understanding? Secondly, is there an easy way to convert a trained auto3dseg algorithm into MONAI bundle format? From what I can see it currently requires manual creation of new metadata.json file etc.

Thanks

**Comment by diazandr3s** [comment]
Hi @mea02jct,

This is great a question. I've also tried to use trained Auto3DSeg models on MONAI Label but it was not a straightforward thing to do.

The bundles created by Auto3DSeg seem to be different from the ones created in the MONAI Model Zoo: https://github.com/Project-MONAI/model-zoo/tree/dev/models - So they can't be consumed by MONAI Label directly.

I had to manually update the [configs](https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/radiology/lib/configs/segmentation.py#L82), [trainers](https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/radiology/lib/trainers/segmentation.py#L73) and [infers](https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/radiology/lib/infers/segmentation.py#L63) files from the Segmentation model in the Radiology app using the hyperparameters created by Auto3DSeg.

I'd suggest opening a ticket or moving this one to the main [MONAI repo](https://github.com/Project-MONAI/MONAI/issues) and see if there is interest from the community in making this easier.

Which models are you training using Auto3DSeg?

Hope this helps,

**Reply by mea02jct** [reply]
Hi Andres,

Thanks for the reply. Good to know that someone else has struggled with this. I shall move my question to the main repo as suggested.

I've currently only trained a cardiac segmentation model, which worked well but hesitating about training more models until I get clarity on which 'standard' MONAI will be using for trained models in future.

Thanks again

**Reply by diazandr3s** [reply]
Hi @mea02jct,

If you already trained the Auto3Seg on cardiac segmentation, I'd suggest you add the model to this Slicer module: https://github.com/lassoan/SlicerMONAIAuto3DSeg

The community will benefit from the cardiac model you've trained.

I am happy to provide more details if you need them.

**Comment by che85** [comment]
Just checking on this issue. Has there been any progress ever since? I am interested in deploying Auto3DSeg, but in a client-server-like way similar to MONAILabel. Auto3DSeg is great, but I wouldn't want the users to download models and run on their local machine which might not be as fast and more cumbersome to maintain.

**Reply by diazandr3s** [reply]
Hi @che85,

I haven't seen much advance on this bundle conversion. However, if your goal is to make the Auto3DSeg models available in MONAI Label for inference only, I created an example showing of the Auto3DSeg haemorrhage model: 

https://drive.google.com/drive/folders/1pzupepOm6A2unCrd9ykxuYKvHAdES1fF?usp=sharing

For this, I changed the configs/segmentation.py and infer/segmentation.py files

This is a quick and simple way of serving those models that can get you going. But I also agree that we may need a more robust way of serving them in MONAI Label via the bundle conversion

Hope this helps,

**Reply by che85** [reply]
@diazandr3s Do you think the Slicer WebServer solution @pieper proposed would be a good solution, or should I go with what you proposed? Currently, if we train, we train separately and would then want to provide the trained models through our local network to our researchers without the need of pulling the models locally.

**Reply by diazandr3s** [reply]
Good question. I haven't tried myself the Slicer WebServer, so I can't comment much on this part. 
I know that modifying the Radiology app in MONAI Label shouldn't be difficult for Auto3DSeg infer purposes :)

**Reply by pieper** [reply]
Making a general framework for hosting Auto3DSeg models in the MONAI Label server would be very cool indeed.  The Slicer WebServer approach should also work and could be easier if you're used to working in Slicer.

**Reply by che85** [reply]
@diazandr3s Does your MONAILabel modification allow multi-volume inputs like the Brain Tumor Segmentation (BRATS)?

**Reply by che85** [reply]
I could imagine refactoring `MONAIAuto3DSegLogic` and extracting the parts responsible for running inference. Similarly to the `monailabel` server, we could create an auto3dseg server that can run independently of Slicer with FastAPI. 

Let me know what you think. I don't want to duplicate work, but `MONAILabel` cannot use multiple input volumes (to the best of my knowledge).

**Reply by che85** [reply]
The Slicer Webserver is also an option and I started experimenting with it.

**Reply by diazandr3s** [reply]
> I could imagine refactoring `MONAIAuto3DSegLogic` and extracting the parts responsible for running inference. Similarly to the `monailabel` server, we could create an auto3dseg server that can run independently of Slicer with FastAPI.
> 
> Let me know what you think. I don't want to duplicate work, but `MONAILabel` cannot use multiple input volumes (to the best of my knowledge).

Hi @che85,

The app I mentioned [here](https://github.com/Project-MONAI/MONAILabel/discussions/1630#discussioncomment-9792600) is an example for single sequence. But with a simple change it should be possible to multisequence inference.

Here I explain how adding a transform that read the other sequence will make this app to run inference on multisequence: 

https://github.com/Project-MONAI/MONAILabel/discussions/1443

and here

https://github.com/Project-MONAI/MONAILabel/issues/241#issuecomment-1497561538 

MONAI Label (server itself) does support training and inference with multisequence. The difficulty comes with the MONAI Label Slicer plugin.

Hope this helps,

**Reply by che85** [reply]
I am working on an implementation adopting Slicer's WebServer. You can host a server from MONAIAuto3DSeg and connect to a remote server. It's a very basic implementation (not session based) but gets the job done.

**Reply by diazandr3s** [reply]
Great! If possible, please share your experience. Thanks in advance!

**Reply by rthug9** [reply]
> I am working on an implementation adopting Slicer's WebServer. You can host a server from MONAIAuto3DSeg and connect to a remote server. It's a very basic implementation (not session based) but gets the job done.

@che85 if possible, could you please explain a bit more on how you did this

**Comment by pieper** [comment]
@che85 if there are no other solutions in the works we could investigate using [Slicer's WebServer](https://slicer.readthedocs.io/en/latest/user_guide/modules/webserver.html) to provide an Auto3DSeg server.  It's already possible to post volumes to Slicer and trigger computation so it wouldn't be too hard to add this feature as part of the extension (or a new extension).
